# Introduction to the Pandas Library 

# What is Pandas and why do we need it?
Pandas is a Python library which offers us a set of built in functions and data structures enabling data analysis and manipulation.

Up to this point, you have learned a number of different data structures. But if you wanted to analyze a dataset that has 100,000 rows of data, storing it in a list, dictionary, or some other basic Python data structure would not be practical.

```
[
  ['TSLA', 483.12, '10-4-2020'],
  ['GOOG', 583.65, '5-26-2020'],
  ['APPL', 802.38, '1-2-2020'],
  ['MSFT', 113.62, '2-6-2020'],
  ['AMZN', 987.96, '3-30-2020'],
  ['TWTR', 53.12, '6-1-2020'],
  ['GOOG', 303.71, '2-14-2020'],
  ['TSLA', 283.65, '3-26-2020'],
  ['AMZN', 583.65, '7-27-2020']
]
```
For example, if you wanted to go through a dataset stored in a list of lists (in the format included above) that included stocks, their prices at multiple dates, and the date at which they were recorded, and then find the average stock price for a given stock, you would have to write the following:

1. Iterate through your list and select every instance where the stock value matches your search string

2. Store those lists in a separate data strucutre

3. Index THOSE lists and select the stock price, and then store those prices in yet another data structure 

4. Write a function to iterate over your list with the prices, return the value, and then map it back to the stock we were initially searching for!


This would be a ton of code, with many places at which an error could pop up and trip us up.


**In Pandas, you can do this in one line of code:**

```
dataframe.groupby('stock_name').mean()
```

You will likely notice that Pandas recreates a lot of functionality you have already learned. The reason why we are okay with using Pandas' approach is because it offers us powerful and easy-to-use tools for manipulating data.

## Upload this file to Colab

- `gdpdata.csv`

### Open this file in Excel and review the data.

In [1]:
from google.colab import files
uploaded = files.upload()

Saving gdpdata.csv to gdpdata.csv


## Read the CSV into Python.

- This is a completely different mechanism than what we've done up to now.
- NOTE.  We're NOT using the `csv` module and `csv.DictReader()` to bring the data into Python.

Instead, we need to `import pandas` and then use 
`read_csv()`.

What data is being created?

In [2]:
# The code in this cell is boilerplate and is just used to import the data and required libraries. 

import pandas as pd

df = pd.read_csv("gdpdata.csv")

print("done")

done


## DataFrames

The core data structure in Pandas is the DataFrame.

The easiest way to think about DataFrames is that they are like spreadsheets. They have column headers specifying a variable, and then rows which represent datapoints.

We will now learn how to explore our data using DataFrames, and in turn we will learn more about what DataFrames are and how we can use them.

# Code-along

When you first instantiate a DataFrame, it is often a good idea to get a sense for the data before you proceed into any more serious analytical work.

One recommended process goes as follows:
1. Check the shape
2. Check the columns
3. Peak at the head
4. Get summary stats for each of the metrics
5. Clean the dataset

In [3]:
# (1) Check the shape
df.shape

(264, 64)

The output from the shape property is a tuple containing the number of rows, followed by the number of columns.
This example data contains 264 rows of data, and 64 columns.

In [4]:
# (2) Check the columns
df.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019'],
      dtype='object')

The output of the columns property is a list of the columns in our DataFrame. As we can see, there are four attributes given to each country followed by a column for each year from 1960 to 2019. Note the type of each column header - this will be important for later if we want to perform manipulations by selecting column names.

In [5]:
# (3) Peak at the head
df.head(5)

Country Name Country Code  ...          2018          2019
0        Aruba          ABW  ...           NaN           NaN
1  Afghanistan          AFG  ...  1.948438e+10  1.910135e+10
2       Angola          AGO  ...  1.013532e+11  9.463542e+10
3      Albania          ALB  ...  1.514702e+10  1.527808e+10
4      Andorra          AND  ...  3.218316e+09  3.154058e+09

[5 rows x 64 columns]

The `head(n)` method takes the number of rows you want to see as an argument, and returns the first `n` rows of the DataFrame. 

What sticks out about this dataset from first looking at the head?

- There are a significant number of `NaNs`
- There appear to be redundant or possibly not useful columns
- The data appears to be sorted alphabetically by country name

In [6]:
# (4) Get summary stats for each of the metrics
df.describe()

1960          1961  ...          2018          2019
count  1.310000e+02  1.320000e+02  ...  2.410000e+02  2.230000e+02
mean   7.106271e+10  7.248501e+10  ...  2.947814e+12  3.237115e+12
std    2.128396e+11  2.213425e+11  ...  9.713264e+12  1.022186e+13
min    1.201203e+07  1.159202e+07  ...  4.258816e+07  4.727146e+07
25%    5.079241e+08  4.891902e+08  ...  1.220712e+10  1.396838e+10
50%    2.760747e+09  2.966849e+09  ...  5.959689e+10  7.608585e+10
75%    2.985372e+10  3.098133e+10  ...  5.554554e+11  7.761167e+11
max    1.369434e+12  1.425105e+12  ...  8.635707e+13  8.769752e+13

[8 rows x 60 columns]

The `describe()` method outputs summary statistics for each row where applicable - giving us a high level understanding of our data. 

What patterns do you notice in the data?

In [7]:
# (5) Clean the dataset

# The process of cleaning data is not as straightforward as the other concepts touched on so far. This module
# will show two simple approaches to dealing with NaNs, though there are other approaches for more advanced 
# programmers that you will be able to tap into once you learn the basics.

# dropna() can be applied to a dataframe and will remove all the rows containing a NaN value. 
# Run this cell to see the output

df.dropna()

Country Name Country Code  ...          2018          2019
11                        Australia          AUS  ...  1.433904e+12  1.392681e+12
12                          Austria          AUT  ...  4.555083e+11  4.463147e+11
14                          Burundi          BDI  ...  3.036932e+09  3.012335e+09
15                          Belgium          BEL  ...  5.426859e+11  5.296067e+11
16                            Benin          BEN  ...  1.425099e+10  1.439071e+10
..                              ...          ...  ...           ...           ...
251  St. Vincent and the Grenadines          VCT  ...  8.113000e+08  8.253852e+08
257                           World          WLD  ...  8.635707e+13  8.769752e+13
261                    South Africa          ZAF  ...  3.682889e+11  3.514316e+11
262                          Zambia          ZMB  ...  2.700524e+10  2.306472e+10
263                        Zimbabwe          ZWE  ...  2.431156e+10  2.144076e+10

[123 rows x 64 columns]

In [9]:
# fillna() can be applied to a DataFrame and will replace all NaN values with whatever argument you pass into the fcn
# Run this cell to see the output

df.fillna(0)

Country Name Country Code  ...          2018          2019
0           Aruba          ABW  ...  0.000000e+00  0.000000e+00
1     Afghanistan          AFG  ...  1.948438e+10  1.910135e+10
2          Angola          AGO  ...  1.013532e+11  9.463542e+10
3         Albania          ALB  ...  1.514702e+10  1.527808e+10
4         Andorra          AND  ...  3.218316e+09  3.154058e+09
..            ...          ...  ...           ...           ...
259        Kosovo          XKX  ...  7.942962e+09  7.926108e+09
260   Yemen, Rep.          YEM  ...  2.759126e+10  0.000000e+00
261  South Africa          ZAF  ...  3.682889e+11  3.514316e+11
262        Zambia          ZMB  ...  2.700524e+10  2.306472e+10
263      Zimbabwe          ZWE  ...  2.431156e+10  2.144076e+10

[264 rows x 64 columns]

## Exercise:

In a separate notebook, do the following:
1. Read in the `GDPData` file as a dataframe
2. Follow each step laid out here 
3. At the bottom of your notebook, include a text cell with the mean GDP in 1963
4. Replace all `NaN` values with a value of your choice

Next, we will begin to manipulate DataFrames. 

Looking at our dataset, it is reasonable to say that there are a few columns that will not be useful to our analysis. 
We will remove the country code, indicator name, and indicator code columns as they will not be necessary for analysis.

The `drop()` function takes a list of strings as an argument, with the list containing each **column** you would like to remove. 

The second argument, `axis`, specifies if you are removing rows or columns. With 0 and 1 respectively representing each option.

In [11]:
df.drop(['Country Code', 'Indicator Name', 'Indicator Code'], axis=1)

Country Name          1960  ...          2018          2019
0           Aruba           NaN  ...           NaN           NaN
1     Afghanistan  5.377778e+08  ...  1.948438e+10  1.910135e+10
2          Angola           NaN  ...  1.013532e+11  9.463542e+10
3         Albania           NaN  ...  1.514702e+10  1.527808e+10
4         Andorra           NaN  ...  3.218316e+09  3.154058e+09
..            ...           ...  ...           ...           ...
259        Kosovo           NaN  ...  7.942962e+09  7.926108e+09
260   Yemen, Rep.           NaN  ...  2.759126e+10           NaN
261  South Africa  7.575397e+09  ...  3.682889e+11  3.514316e+11
262        Zambia  7.130000e+08  ...  2.700524e+10  2.306472e+10
263      Zimbabwe  1.052990e+09  ...  2.431156e+10  2.144076e+10

[264 rows x 61 columns]

In [14]:
#df.columns

#MEAN GDP 1963
#df.describe()
#8.188647e+10

#df.index
#df.mean()

df.fillna("potato")

Country Name Country Code  ...         2018         2019
0           Aruba          ABW  ...       potato       potato
1     Afghanistan          AFG  ...  1.94844e+10  1.91014e+10
2          Angola          AGO  ...  1.01353e+11  9.46354e+10
3         Albania          ALB  ...   1.5147e+10  1.52781e+10
4         Andorra          AND  ...  3.21832e+09  3.15406e+09
..            ...          ...  ...          ...          ...
259        Kosovo          XKX  ...  7.94296e+09  7.92611e+09
260   Yemen, Rep.          YEM  ...  2.75913e+10       potato
261  South Africa          ZAF  ...  3.68289e+11  3.51432e+11
262        Zambia          ZMB  ...  2.70052e+10  2.30647e+10
263      Zimbabwe          ZWE  ...  2.43116e+10  2.14408e+10

[264 rows x 64 columns]

As we can see here, our data cleaning work was not saved from earlier! This is because we didn't update our df variable with the results. 

Let's rerun that below and make sure we save it this time. You can use variables to store DataFrames the same way they can store `lists`, `ints`, or `strings`.

The cell below will assign the updated DataFrame back to the `df` variable.

In [35]:
#df = df.dropna()
#df = df.drop(['Country Code', 'Indicator Name', 'Indicator Code'], axis=1)

#df

## Exercise: 

Update your separate notebook from the earlier exercise to include what we learned about using DataFrame variables.
You should also drop all rows with null values and remove the three columns we removed in this exercise. 

# Moving On:

In Pandas, you can conditionally select rows based on the value a particular column contains. See the example below for selecting the row containing France.

The second example selects two columns and displays them as a DataFrame. 

The third example cell shows how to use iloc to select specific values from your DataFrame. iloc allows you to select based on the row and column index.



In [18]:
df[df['Country Name'] == 'France']

Country Name          1960  ...          2018          2019
75       France  6.222548e+10  ...  2.787864e+12  2.715518e+12

[1 rows x 61 columns]

In [19]:
df[['Country Name','1965']]

Country Name          1965
11                        Australia  2.593684e+10
12                          Austria  9.994071e+09
14                          Burundi  1.589950e+08
15                          Belgium  1.737146e+10
16                            Benin  2.899087e+08
..                              ...           ...
251  St. Vincent and the Grenadines  1.510821e+07
257                           World  1.966264e+12
261                    South Africa  1.133440e+10
262                          Zambia  1.082857e+09
263                        Zimbabwe  1.311436e+09

[123 rows x 2 columns]

You can also iterate over a column in pandas much like how you could iterate through a list. 

In [20]:
for item in df['1965']:
  print(item)

25936835032.0
9994070616.0
158994963.0
17371457608.0
289908720.6
422916848.4
5906636557.0
300392156.9
40069930.07
604377104.4
21790035117.0
45790869.75
150574816.3
54515179581.0
15346741670.0
6026593750.0
70436266147.0
919771356.4
814139855.8
4043901818.0
198318063.9
5760761905.0
592981162.3
2660946061.0
888100000.0
94580860952.0
209370309381.0
224575509177.0
2387048255.0
407845094423.0
24756958695.0
22060164236.0
8589340019.0
147084750.0
101537248148.0
226474285.6
101824755079.0
2053462872.0
7689154053.0
1331399900.0
213235294.1
1550776282591.0
2435078534.0
508650000.0
24406875600.0
353251800.0
401055550813.0
454160665242.0
53617985478.0
17328250389.0
37719212748.0
59554854575.0
2945704143.0
523694949.4
3663333333.0
67978153851.0
972140557.2
90950278258.0
997919320.0
13593932.32
3120833333.0
105189609504.0
118564680313.0
1698319328.0
131174202424.0
432952960394.0
54878902.42
236616225676.0
921600736.3
2948325264.0
833563472.2
21840000000.0
415005604580.0
229455410.9
2956356984.0
79832

In [21]:
df.iloc[4]['1965']

289908720.6

## Exercise: 
    
Select the row which contains South Africa. Use `iloc` to return their GDP in 1974 and 1975. Assign each value to a unique variable.

Create an `if-statement` that checks which year had a larger GDP and prints a statement informing the user of which year is higher.

**CHALLENGE**: Write a `for` loop that iterates through each year of South Africa's GDP data and returns the year with the highest GDP figure, and what that figure is.

In [29]:
#df[df['Country Name'] == 'South Africa']

SA_data = df[['Country Name', '1974', '1975']]
SA_data = SA_data[SA_data['Country Name'] == 'South Africa']

SA_data

#SA_GDP_1974 = df.iloc[4]['1974'] #3.680772e+10
#SA_GDP_1975 = df.iloc[4]['1975'] #3.811454e+10

#print(SA_GDP_1974)

Country Name          1974          1975
261  South Africa  3.680772e+10  3.811454e+10

In [ ]:
#for item in df['1975']:
#  print(item)

In [45]:
SouthAfrica = df[df['Country Name'] == 'South Africa']
SA_First = SouthAfrica.iloc[0]['1974']
SA_Second = SouthAfrica.iloc[0]['1975']
print(SA_First)
print(SA_Second)

if (SA_First > SA_Second):
  print("1974 is higher than 1975")
else:
  print("1975 is higher than 1974")




36807721039.0
38114542813.0
1975 is higher than 1974


In [50]:
print(SouthAfrica)

#Write a for loop that iterates through each year of South Africa's GDP
#data and returns the year with the highest GDP figure, and what that figure is.
rawSA = SouthAfrica.drop(['Country Name'], axis=1)
topItem = 0

for item in rawSA.iloc[0]:
  if item > topItem:
    topItem = item

print(topItem)

#topInfo = SouthAfrica[SouthAfrica[] == topItem]
#print(topInfo)

     Country Name          1960  ...          2018          2019
261  South Africa  7.575397e+09  ...  3.682889e+11  3.514316e+11

[1 rows x 61 columns]
416418874936.0


# Workshop

Now that we have learned how to read in a csv, assess our data, and manipulate it, you will take the ```academy_awards.csv``` file you have been working with and go through the following steps:

1. Read in the data using Pandas' ```read_csv``` function. NOTE: use the following code to import the file, or else you will get an error message: ```df = pd.read_csv("academy_awards.csv", error_bad_lines=False)```
2. Get familiar with your data, jot down some notes in a text cell in your notebook about what stands out
3. ```names = ['Brad Pitt', 'Leonardo Dicaprio', 'Julia Roberts', 'George Clooney']``` Using this list, find the rows in the dataframe where one of these actors were nominated for an award. Write code that iterates over the list of names, printing "In year_nominated, name_of_actor was nominated for an Academy Award." If there are multiple results, choose the second row in your dataframe.

HINT: You will need to use iloc for step 3.


In [30]:
import pandas as pd

#Read in the data using Pandas' read_csv function
awards = pd.read_csv("academy_awards.csv", error_bad_lines=False)

print("done")

done


In [34]:
#awards.shape
#(10137, 11)

#awards.columns
#'Year', 'Category', 'Nominee', 'Additional Info', 'Won?', 'Unnamed: 5',
#'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10'],
#dtype='object'

awards.head(5)

#WHAT STANDS OUT?
#it's weird that there are so many Unnamed headers

Year               Category  ... Unnamed: 9 Unnamed: 10
0  2010 (83rd)  Actor -- Leading Role  ...        NaN         NaN
1  2010 (83rd)  Actor -- Leading Role  ...        NaN         NaN
2  2010 (83rd)  Actor -- Leading Role  ...        NaN         NaN
3  2010 (83rd)  Actor -- Leading Role  ...        NaN         NaN
4  2010 (83rd)  Actor -- Leading Role  ...        NaN         NaN

[5 rows x 11 columns]

In [126]:
names = ['Brad Pitt', 'Leonardo DiCaprio', 'Julia Roberts', 'George Clooney']

#find the rows in the dataframe where one of these actors were nominated for an award
#Write code that iterates over the list of names
#printing "In year_nominated, name_of_actor was nominated for an Academy Award."

#If there are multiple results, choose the second row in your dataframe.

slim_data = awards[['Year', 'Nominee', 'Won?']]
#print(slim_data)

namesOnlydata = []
nominee_Info = {}

for n in names:
  nominee_Info = {}
  namesOnly = slim_data[slim_data['Nominee'] == n]
  nominee_Info[n] = namesOnly['Year']
  namesOnlydata.append(nominee_Info)


#"In year_nominated, name_of_actor was nominated for an Academy Award."
for n in namesOnlydata:
  for key, value in n.items():   
    for item in value:
      item = item[:-7]
      print(f"In {item}, {key} was nominated for an Academy Award.")
    print()

#--------------------------------------


#I know this can't be the right way to do this but I didn't fully understand
#pandas so I just used dictionaries and lists



In 2008, Brad Pitt was nominated for an Academy Award.
In 1995, Brad Pitt was nominated for an Academy Award.

In 2006, Leonardo DiCaprio was nominated for an Academy Award.
In 2004, Leonardo DiCaprio was nominated for an Academy Award.
In 1993, Leonardo DiCaprio was nominated for an Academy Award.

In 2000, Julia Roberts was nominated for an Academy Award.
In 1990, Julia Roberts was nominated for an Academy Award.
In 1989, Julia Roberts was nominated for an Academy Award.

In 2009, George Clooney was nominated for an Academy Award.
In 2007, George Clooney was nominated for an Academy Award.
In 2005, George Clooney was nominated for an Academy Award.

